timesync/ticker.go

```diff
@@ -88,12 +89,11 @@ const sendTickThreshold = 500 * time.Millisecond
 // notify may be skipped also for non-monotonic tick
 // if some of the subscribers where not listening, they are skipped. In that case, errMissedTicks is returned along the number of subscribers not listening
 func (t *Ticker) Notify() (int, error) {
+	t.m.Lock()
```

Comment:
Moving this line here causes the test `TestTicker_StartNotifying` to get into a deadlock. Adding `t.m.Unlock()` inside the next `if` statement fixes this.

However, I'm wondering why move the lock in the first place. Reading the `started` flag unlocked means that in a race we may return an `errNotStarted` here if `Notify` is called right as the ticker is being started. This is not an issue in my opinion as we don't expect a guarantee to get the notified once the `StartNotifying()` is called (best effort / getting notified one layer later should be good enough).

Reply:
> Moving this line here causes the test TestTicker_StartNotifying to get into a deadlock. Adding t.m.Unlock() inside the next if statement fixes this.
Oh yes, I added the unlock in the early return.

> However, I'm wondering why move the lock in the first place. Reading the started flag unlocked means that in a race we may return an errNotStarted here if Notify is called right as the ticker is being started. This is not an issue in my opinion as we don't expect a guarantee to get the notified once the StartNotifying() is called (best effort / getting notified one layer later should be good enough).

I think any data race should be considered as a bug since it's undefined behavior. I see what you mean about that a _raced_ value shouldn't affect the logic in any impactful way, but I think that might be a bit fragile for the future. If that variable is used again in a new place, or for some reason, the data typed gets changed, then any places with that assumption will surprisingly be a problem.

It looks to me that fixing this bug has zero-cost and shouldn't cause any contention to justify having some future risks. However, if you still prefer not moving the lock, I'm happy reverting this part of the PR. But I wouldn't be much convinced :)

Reply:
Fair enough, the cost of obtaining and releasing a lock once per layer is negligible, and I agree that it's more future proof with the protection. 👍🏻